apisix/plugins/loggly.lua

```diff
@@ -0,0 +1,258 @@
+local tab_concat = table.concat
+local udp = ngx.socket.udp
+
+local plugin_name = "loggly"
+local batch_processor_manager = bp_manager_mod.new(plugin_name)
+
+
+local severity = {
+    EMEGR = 0,              --  system is unusable
+    ALERT = 1,              --  action must be taken immediately
+    CRIT = 2,               --  critical conditions
+    ERR = 3,                --  error conditions
+    WARNING = 4,            --  warning conditions
+    NOTICE = 5,             --  normal but significant condition
+    INFO = 6,               --  informational
+    DEBUG = 7,              --  debug-level messages
+}
+
+
+local severity_enums = {}
+do
+    for k, _ in pairs(severity) do
+        severity_enums[#severity_enums+1] = k
+        severity_enums[#severity_enums+1] = k:lower()
+    end
+end
+
+
+local schema = {
+    type = "object",
+    properties = {
+        customer_token = {type = "string"},
+        severity = {
+            type = "string",
+            default = "INFO",
+            enum = severity_enums,
+            description = "base severity log level",
+        },
+        include_req_body = {type = "boolean", default = false},
+        include_resp_body = {type = "boolean", default = false},
+        include_resp_body_expr = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "array"
+            }
+        },
+        tags = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "string",
+                -- we prevent of having `tag=` prefix
+                pattern = "^(?!tag=)[ -~]*",
+            },
+        },
+    },
+    required = {"customer_token"}
+}
+
+
+local defaults = {
+    host = "logs-01.loggly.com",
+    port = 514,
+    protocol = "syslog",
+    timeout = 5000
+}
+
+
+local metadata_schema = {
+    type = "object",
+    properties = {
+        host = {
+            type = "string",
+            default = defaults.host
+        },
+        port = {
+            type = "integer",
+            default = defaults.port
+        },
+        protocol = {
+            type = "string",
+            default = defaults.protocol,
+            -- more methods coming soon
+            enum = {"syslog"}
+        },
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            default= defaults.timeout
+        },
+        log_format = {
+            type = "object",
+        }
+    }
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 411,
+    name = plugin_name,
+    schema = batch_processor_manager:wrap_schema(schema),
+    metadata_schema = metadata_schema
+}
+
+
+function _M.check_schema(conf, schema_type)
+    if schema_type == core.schema.TYPE_METADATA then
+        return core.schema.check(metadata_schema, conf)
+    end
+
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return nil, err
+    end
+    return log_util.check_log_schema(conf)
```

Comment:
```suggestion
    if schema_type == core.schema.TYPE_METADATA then
        return core.schema.check(metadata_schema, conf)
    end
    return core.schema.check(schema, conf)
```

can be handled in this way.

Reply:
Hi, thanks for the review. As the plugin schema contains, `include_resp_body_expr`, this need to be validated. core.schema.check just performs jsonschema validation based on the constraints set to the schema.  
https://github.com/apache/apisix/blob/a813b5aa8980dd8fa5780af351b6b47b5219d0b9/apisix/core/schema.lua#L31-L65

while the check_log_schema performs validation on the custom DSL
https://github.com/apache/apisix/blob/a813b5aa8980dd8fa5780af351b6b47b5219d0b9/apisix/utils/log-util.lua#L214-L228

In this case, we need to perform validation on log_schema. Thanks.

Reply:
OK, understood